File: 2011/J112.pas
var
  n, m, x, y, i: longint;
  count, twos, threes, ans: longint;
begin
  read(n, m, x);
  count := 1;
  twos := 0;
  threes := 0;
  for i := 1 to n - 1 do
  begin
    read(x, y);
    if (x = y) then
      inc(count)
    else
      count := 1;
    if (count mod 3 = 0) then
    begin
      dec(twos);
      inc(threes)
    end;
    if (count mod 3 = 2) then
      inc(twos)
  end;
  ans := threes * 2 + twos;
  if (ans + m > n) then
    ans := n - m;
  writeln(ans)
end.

File: 2011/J112.cpp
#include <cstdio>
int main() {
  int n, m, x, y, count = 1;
  int twos = 0, threes = 0;
  scanf("%d %d %d", &n, &m, &x);
  for (int i = 0; i < n - 1; i++) {
    scanf("%d %d", &x, &y);
    if (x == y) {
      count++;
    } else {
      count = 1;
    }
    if (count % 3 == 0) {
      twos--;
      threes++;
    }
    if (count % 3 == 2) {
      twos++;
    }
  }
  int ans;
  if (m >= threes + twos) {
    ans = threes * 2 + twos;
  } else if (m <= threes) {
    ans = m * 2;
  } else {
    ans = threes + m;
  }
  if (ans + m > n) {
    ans = n - m;
  }
  printf("%d\n", ans);
  return 0;
}
